polkadot/runtime/common/src/xcm_sender.rs

```diff
@@ -119,7 +119,9 @@ where
 		let config = configuration::ActiveConfig::<T>::get();
 		let para = id.into();
 		let price = P::price_for_delivery(para, &xcm);
-		let blob = W::wrap_version(&d, xcm).map_err(|()| DestinationUnsupported)?.encode();
+		let versioned_xcm = W::wrap_version(&d, xcm).map_err(|()| DestinationUnsupported)?;
+		versioned_xcm.validate_xcm_nesting().map_err(|()| ExceedsMaxMessageSize)?;
+		let blob = versioned_xcm.encode();
```

Comment:
So why is it being added to the router if having it in the pallet is better? Just for validation to return an error earlier?

Reply:
> ...if having it in the pallet is better

I don't quite understand what you mean by this, could you elaborate more? Or what do you suggest here?

---
You added `validate_xcm_nesting()` to the XcmpQueue for `SendXcm::validate`, and I just added it exactly like that to the `ParentAsUmp` and `ChildParachainRouter`.
Thank you for pointing me to the that sr_labs issue, now I see it was really about HRMP. So, if we don't have this issue for UMP/DMP, we can close this PR.

My point/question is that, without this PR, we don't allow "too-much-nested" XCM between paras, why do we allow to "too-much-nested" XCM from para to relaychain or from relaychain to para? Again, if it is ok, we can close this PR :)

Also there are other PRs which allows/relaxes XCM execution e.g.: https://github.com/polkadot-fellows/runtimes/pull/261, so maybe better safe then sorry?

E.g. can we have a scenario like this? Like the relaychain allows to execute XCM which send "too-much-nested" XCM to the parachain (**DMP works**), which would result re-sending the "too-much-nested" XCM to the parachain (**HRMP fails**)? Is that ok?


@ggwpez what do you suggest with this PR? :)

Reply:
>Like the relaychain allows to execute XCM which send "too-much-nested" XCM to the parachain

"too-much-nested" <=> fails-to-decode.  
An XCM that is too deeply nested will fail to `Decode::decode`. It is undecodable, that is the whole issue.  
I think we should also not send such messages over any other protocol, including DMP/UMP. I just did not think about this when writing it for HRMP.